src/Solnet.Rpc/Utilities/RateLimiter.cs

```diff
@@ -0,0 +1,134 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Solnet.Rpc.Utilities
+{
+    /// <summary>
+    /// A primitive blocking sliding time window rate limiter. Not thread-safe.
+    /// </summary>
+    public class RateLimiter : IRateLimiter
+    {
+        private int _hits;
+        private int _duration_ms;
+        private Queue<DateTime> _hit_list;
+
+        /// <summary>
+        /// Create a simple rate-limit tracking instance.
+        /// This allows a number of hits within a window of duration_ms. 
+        /// </summary>
+        /// <param name="hits">Number of allowed hits</param>
+        /// <param name="duration_ms">Duration of timespan in miliseconds</param>
+        public RateLimiter(int hits, int duration_ms)
+        {
+            _hits = hits;
+            _duration_ms = duration_ms;
+            _hit_list = new Queue<DateTime>();
+        }
+
+        /// <summary>
+        /// Create a simple rate-limit tracking instance.
+        /// Configure using `PerSeconds` and `AllowHits` 
+        /// </summary>
+        public static RateLimiter Create()
+        {
+            return new RateLimiter(1, 0);
+        }
+
+        /// <summary>
+        /// Would a fire request be allowed without delay?
+        /// </summary>
+        /// <returns></returns>
+        public bool CanFire()
+        {
+            var checkTime = DateTime.UtcNow;
+            var resumeTime = NextFireAllowed(checkTime);
+            return checkTime >= resumeTime;
+        }
+
+        /// <summary>
+        /// Pre-fire check - this will block if fire rates exceed defined limits until valid.
+        /// </summary>
+        public void Fire()
+        {
+
+            var checkTime = DateTime.UtcNow;
+            var resumeTime = NextFireAllowed(checkTime);
+            while (DateTime.UtcNow <= resumeTime)
+                Thread.Sleep(50);
+
+            // record this trigger
+            _hit_list.Enqueue(DateTime.UtcNow);
+
+        }
+
+        /// <summary>
+        /// When is a next fire allowed?
+        /// </summary>
+        /// <param name="checkTime"></param>
+        /// <returns></returns>
+        private DateTime NextFireAllowed(DateTime checkTime)
+        {
+            DateTime resumeTime = checkTime;
+
+            // sliding window not set, allow everything through immediately
+            if (_duration_ms == 0) return resumeTime;
```

Comment:
If `_duration_ms == 0` we have a leak!

On `Fire()` you Enqueue the values regardless, and here you exit early. This means the Queue is going to grow indefinitely  (specially worrisome for the default value used when no rate limiter is provided!)

Reply:
Good spot. Added _duration_ms>0 guard to enqueue.
Also dropped duration_ms=0 default rate limiter for RPC in favour of your null rate limiter suggestion.